Summer of code summary for 2023 edition (#492)
apple/swift-org-website: _posts/2024-01-25-summer-of-code-2023-summary.md

---
layout: post
published: true

> date: 2022-12-05 10:00:00

Author: Done thanks!

> title: Swift Google Summer of Code 2023 Summary

Author: Okey

> author: [ktoso]

Author: Todo: maybe we can add all mentors here as coauthors?

---

> The Swift project regularly participates in Google Summer of Code in order to help people new to the open source ecosystem dip their toes in contributing to Swift and its wide ecosystem.

Author: good call, linked!

During the 2023 edition of the program, we ran three projects, all of which completed their assigned projects successfully. We would like to thank all the contributors for the time and passion they poured into their projects!

The projects in this edition were:

* Swift Memcache library
* Incremental Parsing in SwiftParser
* KeyPath inference and diagnostic improvements

> We’d like to extend our sincere thanks to the participants and mentors for completing these awesome projects, and use this post to highlight their work to the wider community. Below, each project is described in a small summary.

Author: Done thanks!

> We are always looking for ideas, mentors, and general input about how we can make Swift’s ecosystem more inclusive and easier to participate in. If you have some ideas, would like to become a mentor in future editions, or if you’re just curious about other project ideas that were floated during previous editions, visit the [dedicated GSoC category](https://forums.swift.org/c/development/gsoc/98) on the Swift forums. You can also check out [last year’s GSoC summary blog post](https://www.swift.org/blog/swift-summer-of-code-2022-summary/), where we highlighted last year’s projects. 

Author: Very true, fits well at the end without rewording as well -- done

If you’re interested in participating as a Mentee, keep an eye on the official [GSoC schedule](https://summerofcode.withgoogle.com/) and on the Swift forum’s [GSoC category](https://forums.swift.org/c/development/gsoc/98).

> What follows are short descriptions, written in collaboration between mentees and mentors, summarizing this year’s projects as well as a few impressions about the overall program:

Author: Sounds nice, I've taken that sentence

## Project notes 

### Swift Memcache

* Mentee: [Delkhaz Ibrahimi](https://github.com/dkz2)
* Mentor: [Franz Busch](https://github.com/FranzBusch)

> The goal of the project was to develop a native Memcache connection abstraction for the Swift on Server ecosystem. This connection should be implemented using `SwiftNIO,` offer native Swift Concurrency APIs and integrate with the rest of the server ecosystem. The focus during the project was implementing the Memcache [meta command protocol](https://github.com/memcached/memcached/wiki/MetaCommands) and offering basic `get` and `set` functionalities. Below is a short example how the new `MemcacheConnection` type can be used to `set` and `get` a value for a given key:

Author: Adjusted wording taking all these into account, thank you!

let memcacheConnection = MemcacheConnection(host: "127.0.0.1", port: 11211, eventLoopGroup: .singleton)

// Initialize the service group
let serviceGroup = ServiceGroup(services: [memcacheConnection], logger: logger)

try await withThrowingTaskGroup(of: Void.self) { group in
    group.addTask { try await serviceGroup.run() }
    
    // Set a value for a key.
    let setValue = "bar"
    try await memcacheConnection.set("foo", value: setValue)
    
    // Get the value for a key.
    // Specify the expected type for the value returned from Memcache.
    let getValue = try await memcacheConnection.get("foo", as: String.self)
}
```

After finishing the foundational work to implement the `get` and `set` commands, support for `delete`, `append`, `prepend`, `increment` and `decrement` was added. Lastly, support for checking and updating the time-to-live for keys was added.

The new `MemcacheConnection` type lays the ground work to implement a higher level `MemcacheClient` that offers additional functionality such as connection pooling, retries, key distribution across nodes and more. However, implementing such a client was out of scope for this year’s GSoC project.

> If you’d like to learn more about this project and Delkhaz’s experience, head over to this summary post in the summary post of the  [[GSoC 2023] swift-memcache-gsoc Project Kickoff ****](https://forums.swift.org/t/gsoc-2023-swift-memcache-gsoc-project-kickoff/64932)Swift forums thread.

Author: The last post in this thread is the "summary" I'll reword the link, thank you!

### Implement Incremental Re-Parsing in SwiftParser

* Mentee: [Ziyang Huang](https://github.com/StevenWong12) 
* Mentor: [Alex Hoppen](https://github.com/ahoppen)

> The project improved the performance of SwiftParser in editor contexts like syntax highlighting where minor edits are applied to the file, by incrementally parsing the source files and re-using parts of the syntax tree that remain unchanged. At the same time, performance loss during normal parsing, for example in the context of the compiler, was kept to a minimum.

Author: > SwiftParser

It's part of SwiftSyntax package -- but I can link to something that is a good entrypoint for it 👍

> The project improved the performance of SwiftParser in editor contexts like syntax highlighting where minor edits are applied to the file, by incrementally parsing the source files and re-using parts of the syntax tree that remain unchanged. At the same time, performance loss during normal parsing, for example in the context of the compiler, was kept to a minimum.

Author: I think that rewording is right, Applied

> The project improved the performance of SwiftParser in editor contexts like syntax highlighting where minor edits are applied to the file, by incrementally parsing the source files and re-using parts of the syntax tree that remain unchanged. At the same time, performance loss during normal parsing, for example in the context of the compiler, was kept to a minimum.

Reviewer: If you wanted to link to an entry point, I would choose https://github.com/apple/swift-syntax/tree/main/Sources/SwiftParser

> The project improved performance of SwiftParser in editor contexts like syntax highlighting where minor edits are applied to the file, by incrementally parsing the source files and re-using parts of the syntax tree that remain unchanged. At the same time it made sure that it did not incur too much performance loss when incremental parsing is not used, for example in the context of the compiler.

Reviewer 2: Maybe something like this reads a bit better?

```suggestion
The project improved the performance of SwiftParser in editor contexts like syntax highlighting where minor edits are applied to the file, by incrementally parsing the source files and re-using parts of the syntax tree that remain unchanged. At the same time, performance loss during normal parsing, for example in the context of the compiler, was kept to a minimum.
```

The most challenging part of this project was to make sure we parse the source files correctly. Considering the code snippet below:

```swift
foo() {}

> someLabel: ⏩️switch x⏪️ {

Reviewer 3: I'd recommend removing the emojis and simply saying, "When we remove the `switch x` expression, a …"

Reviewer: Sounds like a good idea. I think the emojis are a leftover of how we marked up the tests 😄

Author: Oh that's fun, didn't know that's from tests.

I removed and reworded below :)

> The most challenging part of this project is that we need to make sure we parse the source files correctly. Considering the code snippet below:

Reviewer 2: We should use past tense as the project is already finished.

```suggestion
The most challenging part of this project was to make sure we parse the source files correctly. Considering the code snippet below:
```

    default: break
}
```

This source is parsed as a `FunctionCallExprSyntax` and a `LabeledStmtSyntax`. When we remove the code between ⏩️ and ⏪️, a naïve implementation could reuse `foo() {}` as a function call since the edit didn’t touch it. But this is not correct since the `someLabel` block now became a labeled trailing closure of `foo() {}`.

> To solve this problem, we collect some additional information for every syntax node during the initial parse to mark the possibly influence range for each node. That information is used to correctly re-parse the `foo()` function call and include `someLabel` as a labeled trailing closure in the call.

Reviewer 3: "the potentially affected range for each node" ?

> The implementation speeds up parsing by about **10x** when we parse source incrementally while **only incurring a 2~3% of performance loss** during normal parsing. 

Reviewer 3: Nice!

Author: :-)

If you’d like to read more about this project and Ziyang’s experience in his own words, head over to [his GSoC experience post](https://forums.swift.org/t/gsoc-2023-my-gsoc-experience/67340) on the Swift forums.

> ### Key Path inference and diagnostic improvements

Reviewer 3: Capitalize: "Key Path Inference and Diagnostic Improvements" to match other blog posts

Author: gotcha

Author: fixed above as well

> If you’d like to read more about this project and Ziyang’s experience his own words, head over to this post on the Swift forums: [[GSoC 2023] My GSoC Experience](https://forums.swift.org/t/gsoc-2023-my-gsoc-experience/67340).

Reviewer 2: Same feedback here as for the other link. e.g.

```suggestion
If you’d like to read more about this project and Ziyang’s experience in his own words, head over to [his GSoC experience post](https://forums.swift.org/t/gsoc-2023-my-gsoc-experience/67340) on the Swift forums.
```

* Mentee: [Amritpan Kaur](https://github.com/amritpan)
* Mentor: [Pavel Yaskevich](https://github.com/xedin?tab=repositories) 

This project was focused on performance and diagnostic improvements to type-checking of key path literal expressions as well as improvements to new features such as keypaths-as-functions introduced to the language by [SE-0249](https://github.com/apple/swift-evolution/blob/main/proposals/0249-key-path-literal-function-expressions.md).

> During compilation, the key path expression root and value were type-checked sequentially to resolve a key path type from this context. However, the design of how the type-checker evaluates key path component types, their relationships to each other, and key path capabilities results in hard to understand compiler errors and even failures to type-check some valid Swift code. 

Reviewer 3: "However, the design of the type-checker's evaluation of key path component types, their …"

Some of the problems with this approach can be illustrated by the following code example:

```swift
struct User {
  var name: Name
}
struct Name {
  let firstName: String
}

func test(_: WritableKeyPath<User, String>) {}

test(\.name.firstName)
```

The compiler produces the following error:

```swift
error: key path value type 'WritableKeyPath<User, String>' cannot be converted to contextual type 'KeyPath<User, String>'
test(\.name.firstName)
     ^
```

> There are multiple issues with this error diagnostic - contextual type is actually `WritableKeyPath`  and key path should be inferred as read-only, the source information is lost and compiler is enable to point out a problem with an argument to a call to the function `test`.

Reviewer 3: "with this error diagnostic: contextual …"

Reviewer 3: Should it be "compiler is unable to point out" ?

Author: Thanks for spotting, yes

> To address these and other issues we explored a different design for key path literal expression type-checking: infer the root type of the key path first and propagate that information to the components and infer capability based on the components before setting the type for the key path expression. This made it a lot easier to diagnose contextual type failures and support previously failing conversions. This approach improves the performance as well because literal is resolved only if the context expects a key path and root type is either provided by the developer explicitly or is able to be inferred from the context. 

Reviewer 3: "To address these and other issues, we explored …"

Reviewer 3: "because the literal is resolved only"

> To address these and other issues we explored a different design for key path literal expression type-checking: infer the root type of the key path first and propagate that information to the components and infer capability based on the components before setting the type for the key path expression which made it a lot easier to diagnose contextual type failures and support previously failing conversions. This approach improves the performance as well because literal is resolved only if context expects a key path and root type is either provided by the developer explicitly or is inferable from the context. 

Reviewer 2: This paragraph is quite difficult to read at the moment. Maybe we can split it up into shorter sentences?

Author: I broke it up a bit, thanks!

> With new approach implemented the compiler now produces the following diagnostic:

Reviewer 2: ```suggestion
With the new approach implemented the compiler now produces the following diagnostic:
```

```swift
error: cannot convert value of type 'KeyPath<User, String>' to expected argument type 'WritableKeyPath<User, String>'
test(\.name.firstName)
     ^
```

> If you’d like to learn more about the details of this project we recommend having a look at this excellent and [very detailed writeup written by Amritpan](https://forums.swift.org/t/key-path-inference-and-diagnostic-improvements-an-update/69632) on the forums.

Reviewer 3: "of this project, we recommend"

> ## Mentee impressions

Reviewer 3: Capitalize Impressions

> If you’d like to learn more about the details of this project we recommend having a look at this excellend and [very detailed writeup written by Amritpan](https://docs.google.com/document/d/19hWKez1sx7W5XMDc9sh7fRDAlPdMP2FxYuIyGs7MCbI/edit).

Reviewer 4: I've published this on the Swift Forum just now and this can be linked as - [[Key Path inference and diagnostic improvements](https://forums.swift.org/t/key-path-inference-and-diagnostic-improvements-an-update/69632)](https://forums.swift.org/t/key-path-inference-and-diagnostic-improvements-an-update/69632)

Author: Awesome, thank you @amritpan ! :-)

> If you’d like to learn more about the details of this project we recommend having a look at this excellend and [very detailed writeup written by Amritpan](https://forums.swift.org/t/key-path-inference-and-diagnostic-improvements-an-update/69632) on the forums.

Reviewer 2: ```suggestion
If you’d like to learn more about the details of this project we recommend having a look at this excellent and [very detailed writeup written by Amritpan](https://forums.swift.org/t/key-path-inference-and-diagnostic-improvements-an-update/69632) on the forums.
```

> It’s great to see what projects our GSoC participants accomplished this year! Even better though is seeing participants enjoy their time working with the Swift project, as a big part of Summer of Code is giving mentees (and mentors) a space to grow and learn about Open Source development. 

Reviewer 3: Lowercase "open source"

> Here’s a few impressions of the mentees about their experience: 

Reviewer 3: "Here are a few …"

**Ziyang (Implement Incremental Re-Parsing in SwiftParser):**

> After measurement, our implementation speed up parsing about 10 times when we parse source incrementally while only bring 2~3% of performance loss to normal parsing. 🎉 🎉
> 
> I also bring this feature to sourcekit-lsp and swift-stress-tester, it is really exciting to see my work can be actually put into use.
> 
> Special thanks to my mentor Alex for the quick response, detailed reviews and inspiring ideas.

**Amritpan (Key Path inference and diagnostic improvements):**

> I enjoyed working on this project this year, primarily because it was a challenge that allowed me to deepen my understanding of the solver implementation, utilize improvements I made last year to the debug output, and make more impactful changes to the compiler codebase. 
> 

> > Refactoring the debug output last year helped me understand how the various pieces of information that the type checker collected were then evaluated to assign types from context. Taking a look at key path expression type checking failures revealed some of the fallibilites of the constraint system and solver and how various design decisions choices could solve certain issues while causing others.

Reviewer 2: ```suggestion
> Refactoring the debug output last year helped me understand how the various pieces of information that the type checker collected were then evaluated to assign types from context. Taking a look at key path expression type checking failures revealed some of the fallibilities of the constraint system and solver and how various design decision choices could solve certain issues while causing others.
```

> 

> We hope that these notes inspire you to apply to work with us on the Swift project in future Google Summer of Code editions!

Reviewer 3: Recommend rewording this a little because not everyone is eligible. How about: "We hope this writeup inspires you to either apply to work with us or become a mentor on the Swift project in a future …"

Author: I like this, thank you
